src/framework/accessibility/internal/accessibilitycontroller.cpp

```diff
@@ -0,0 +1,205 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "accessibilitycontroller.h"
+
+#include <QGuiApplication>
+#include <QAccessible>
+
+#include "accessibleobject.h"
+#include "accessiblecontrollerinterface.h"
+#include "accessibleiteminterface.h"
+
+#include "async/async.h"
+#include "log.h"
+
+using namespace mu::accessibility;
+
+AccessibilityController::~AccessibilityController()
+{
+}
+
+static QAccessibleInterface* muAccessibleFactory(const QString& classname, QObject* object)
+{
+    if (classname == QLatin1String("mu::accessibility::AccessibleObject")) {
+        return static_cast<QAccessibleInterface*>(new AccessibleItemInterface(object));
+    }
+
+    if (classname == QLatin1String("mu::accessibility::AccessibilityController")) {
+        return static_cast<QAccessibleInterface*>(new AccessibleControllerInterface(object));
+    }
+
+    return nullptr;
+}
+
+static void updateHandlerNoop(QAccessibleEvent*)
+{
+}
+
+void AccessibilityController::init()
+{
+    setObjectName("AccessibilityController");
+
+    QAccessible::installFactory(muAccessibleFactory);
+
+    QAccessible::setRootObject(this);
+    QAccessible::installRootObjectHandler([](QObject*) {});
+
+    //! NOTE Disabled any events from Qt
+    QAccessible::installUpdateHandler(updateHandlerNoop);
+}
+
+void AccessibilityController::created(IAccessibility* parent, IAccessibility* item)
+{
+    //! TODO Not working yet
+    parent = nullptr;
+
+    QObject* prnObj = nullptr;
+    if (parent) {
+        prnObj = findItem(parent).object;
+    } else {
+        prnObj = this;
+    }
+
+    Item it;
+    it.parent = parent;
+    it.item = item;
+    it.object = new AccessibleObject(item, prnObj);
+    it.object->setController(shared_from_this());
+    it.iface = QAccessible::queryAccessibleInterface(it.object);
+
+    m_items.append(it);
+
+    LOGI() << "parent: " << (parent ? parent->accessibleName() : "") << ", item: " << item->accessibleName();
+
+    QAccessibleEvent ev(it.object, QAccessible::ObjectCreated);
+    sendEvent(&ev);
+}
+
+void AccessibilityController::destroyed(IAccessibility* aitem)
+{
+    LOGI() << aitem->accessibleName();
+    int idx = indexBy(aitem);
+    IF_ASSERT_FAILED(idx >= 0) {
+        return;
+    }
+    Item item = m_items.takeAt(idx);
+    QAccessibleEvent ev(item.object, QAccessible::ObjectDestroyed);
+    sendEvent(&ev);
+
+    delete item.object;
```

Comment:
do we really need to delete object here? The point is that it's a bit unclear when this function would be called (before an object will be destroyed or after)

Perhaps make sense to rename the function

Reply:
I'm also thinking about the existing QObject destruction mechanism. How likely it would cross-over with this functionality

Reply:
This is a service object, which the controller created itself, so the controller deleted this object then it not need more.

Reply:
> I'm also thinking about the existing QObject destruction mechanism. How likely it would cross-over with this functionality

For this, need set parent to object, in this case, controller creating this object, so it is parent, but! controller live all time, so all objects will be live all time.

Reply:
got it